### PR TITLE
refactor: use a shared source set for all apple targets

### PR DESCRIPTION
**Pull request by typfel**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Currently we only support building Kalium for the iOS x86 simulator.

### Solutions

Support building for all iOS targets & macOS target with a shared source set called `darwin`.

See: https://kotlinlang.org/docs/multiplatform-share-on-platforms.html#share-code-in-libraries

There's a lot code duplication going on in the Gradle files. I'm interesting in pointers on how to reduce it.

### Notes

 I disabled the Carthage plugin in the `cryptography` module because it's hard coded for the simulator and I don't want to update it since we won't need it when CoreCrypto becomes a multiplatform library.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: buildSrc/src/main/kotlin/com/wire/kalium/plugins/Multiplatform.kt [hunk 1]
```diff
@@ -37,6 +37,9 @@ fun Project.configureDefaultMultiplatform(
             //       this should require us moving iOS code from iOSX64Main to
             //       another sourceSet (iOSMain/iOSCommon or similar)
             iosX64()
+            iosArm64()
+            iosSimulatorArm64()
+            macosX64()
```
Comment (vitorhugods) [inline buildSrc/src/main/kotlin/com/wire/kalium/plugins/Multiplatform.kt:42]: I don't know why the `getting` and `creating` delegates are not working in this file, at least the IDE doesn't like it but it seems to work.

We can get away by not using the sugar syntax and going a bit more primitive here:

```suggestion
            macosX64()
            
            val commonMain = sourceSets.getByName("commonMain")
            val commonTest = sourceSets.getByName("commonTest")
            val darwinMain = sourceSets.create("darwinMain") {
                dependsOn(commonMain)
            }
            val darwinTest = sourceSets.create("darwinTest") {
                dependsOn(commonTest)
            }
            val iosX64Main = sourceSets.getByName("iosX64Main") {
                dependsOn(darwinMain)
            }
            val iosX64Test = sourceSets.getByName("iosX64Test") {
                dependsOn(darwinTest)
            }
            val iosArm64Main = sourceSets.getByName("iosArm64Main") {
                dependsOn(darwinMain)
            }
            val iosArm64Test = sourceSets.getByName("iosArm64Test") {
                dependsOn(darwinTest)
            }
            val iosSimulatorArm64Main = sourceSets.getByName("iosSimulatorArm64Main") {
                dependsOn(darwinMain)
            }
            val iosSimulatorArm64Test = sourceSets.getByName("iosSimulatorArm64Test") {
                dependsOn(darwinTest)
            }
            val macosX64Main = sourceSets.getByName("macosX64Main") {
                dependsOn(darwinMain)
            }
            val macosX64Test = sourceSets.getByName("macosX64Test") {
                dependsOn(darwinTest)
            }
```

Maybe extracting this block to a function would be better:

```kotlin
fun KotlinMultiplatformExtension.darwinMultiplatformConfig() {
    val commonMain = sourceSets.getByName("commonMain")
    val commonTest = sourceSets.getByName("commonTest")
    val darwinMain = sourceSets.create("darwinMain") {
        dependsOn(commonMain)
    }
    val darwinTest = sourceSets.create("darwinTest") {
        dependsOn(commonTest)
    }
    val iosX64Main = sourceSets.getByName("iosX64Main") {
        dependsOn(darwinMain)
    }
    val iosX64Test = sourceSets.getByName("iosX64Test") {
        dependsOn(darwinTest)
    }
    val iosArm64Main = sourceSets.getByName("iosArm64Main") {
        dependsOn(darwinMain)
    }
    val iosArm64Test = sourceSets.getByName("iosArm64Test") {
        dependsOn(darwinTest)
    }
    val iosSimulatorArm64Main = sourceSets.getByName("iosSimulatorArm64Main") {
        dependsOn(darwinMain)
    }
    val iosSimulatorArm64Test = sourceSets.getByName("iosSimulatorArm64Test") {
        dependsOn(darwinTest)
    }
    val macosX64Main = sourceSets.getByName("macosX64Main") {
        dependsOn(darwinMain)
    }
    val macosX64Test = sourceSets.getByName("macosX64Test") {
        dependsOn(darwinTest)
    }
}

```

**Review by vitorhugods — CHANGES_REQUESTED** [review]
👨‍🍳👌

Great effort. Looking forward to see Kalium running _everywhere_ soon :D 

Just one suggestion to the copy/pasting problem.

File: buildSrc/src/main/kotlin/com/wire/kalium/plugins/Multiplatform.kt (hunk 1 above)
Comment (typfel) [reply buildSrc/src/main/kotlin/com/wire/kalium/plugins/Multiplatform.kt]: Updated

**Review by vitorhugods — APPROVED** [review]
😍

**Pull request merged**